### DISCUSSION TITLE
Calls build() each time I call ref.read method

### DISCUSSION THREAD

**Discussion by dyardy** [rrousselGit/riverpod#2350] [General]
I have the following Provider, which maintains the user position (lat and long).  Within my user interface, this is called when the position changes.

` ref.read(positionNotifierProvider.notifier).updateLocation(MyWeatherPosition(position.latitude, position.longitude));      `

The problem is every time this updateLocation is called, the build() is called, setting the state back to its original position.  I expected this would only be called once if the user stayed on that page.

```
  @override
  MyWeatherPosition build() {
    return MyWeatherPosition(39.327787, -84.379258);
  }
```

As a work-around used @Riverpod(keepAlive: true) (which ensures that build is only called once) but this is not preferred.

Am I doing something incorrect?

```
@riverpod
class PositionNotifier extends _$PositionNotifier {
  
  @override
  MyWeatherPosition build() {
    return MyWeatherPosition(39.327787, -84.379258);
  }

  void updateLocation(MyWeatherPosition positionItem) {
    if (isDifferent(positionItem)) {
      state = positionItem;
       ref.read(positionItemClassProvider.notifier).Add(positionItem);
    }
  }

  bool isDifferent(MyWeatherPosition newPosition) {
    if (state.lat != newPosition.lat || state.lon != newPosition.lon) {
      return true;
    }

    return false;
  }
}
```

and within my ConsumerState class

```
void _updatePositionList(PositionItemType type, String displayValue, Position? position) {
    if (position != null) {
      ref
          .read(positionNotifierProvider.notifier)
          .updateLocation(MyWeatherPosition(position.latitude, position.longitude));      
    }
  }
```

**Comment by dyardy** [comment]
As a check I used the old approach and didn't have any trouble.

I thought my @riverpod annotation was equivalent to this below.



For completeness, I used this to call the update
```
   ref
          .read(positionNotifier2Provider.notifier)
          .updateLocation(MyWeatherPosition(position.latitude, position.longitude));
```



```
class PositionNotifier2 extends StateNotifier<MyWeatherPosition> {
  PositionNotifier2(this.ref) : super(MyWeatherPosition(39.327787, -84.379258));

  Ref ref;

  void updateLocation(MyWeatherPosition positionItem) {
    if (isDifferent(positionItem)) {
      state = positionItem;
      ref.read(positionItemClassProvider.notifier).Add(positionItem);
    }
  }

  bool isDifferent(MyWeatherPosition newPosition) {
    if (state.lat != newPosition.lat || state.lon != newPosition.lon) {
      return true;
    }

    return false;
  }
}

final positionNotifier2Provider = StateNotifierProvider<PositionNotifier2, MyWeatherPosition>((ref) {
  return PositionNotifier2(ref);
});
```

**Reply by motucraft** [reply]
This is not equivalent to @riverpod because StateNotifierProvider is not StateNotifierProvider.autoDispose, right?
If you don't add .autoDispose, I think it is equivalent to @Riverpod(keepAlive: true).

**Reply by dyardy** [reply]
I found a clue which resolved the problem.  By default, the provider will start up and be available on the page (without newing up again) if there is a watch on that page.  Without the specific watch on the page, the provider would start fresh each time.
